### PR TITLE
[Refactor] Refactor ema hook

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tests/test_hooks/test_ema_hook.py [hunk 1]
```diff
@@ -62,239 +39,237 @@ class ToyModel3(ToyModel):
...
+        for key in ori_checkpoint['state_dict'].keys():
+            assert_allclose(
+                ori_checkpoint['state_dict'][key].cpu(),
+                checkpoint['ema_state_dict'][f'module.{key}'].cpu())
+
+            assert_allclose(
+                ema_hook.ema_model.state_dict()[f'module.{key}'].cpu(),
+                checkpoint['state_dict'][key].cpu())
+
+    def test_after_load_checkpoint(self):
```
Comment (RangiLyu) [inline tests/test_hooks/test_ema_hook.py:182]: add test case for resume and not resume

File: mmengine/hooks/ema_hook.py [hunk 2]
```diff
@@ -50,9 +50,9 @@ def __init__(self,
         assert not (begin_iter != 0 and begin_epoch != 0), (
             '`begin_iter` and `begin_epoch` should not be both set.')
         assert begin_iter >= 0, (
-            f'begin_iter must larger than 0, but got begin: {begin_iter}')
+            f'`begin_iter` must larger than 0, but got begin: {begin_iter}')
```
Comment (zhouzaida) [inline mmengine/hooks/ema_hook.py:53]:
```suggestion
            f'`begin_iter` must be larger than or equal to 0, but got begin: {begin_iter}')
```

File: mmengine/hooks/ema_hook.py [hunk 3]
```diff
@@ -50,9 +50,9 @@ def __init__(self,
...
         assert begin_epoch >= 0, (
-            f'begin_epoch must larger than 0, but got begin: {begin_epoch}')
+            f'`begin_epoch` must larger than 0, but got begin: {begin_epoch}')
```
Comment (zhouzaida) [inline mmengine/hooks/ema_hook.py:55]:
```suggestion
            f'`begin_epoch` must be larger than or equal to 0, but got begin: {begin_epoch}')
```

File: tests/test_hooks/test_ema_hook.py [hunk 4]
```diff
@@ -62,239 +39,249 @@ class ToyModel3(ToyModel):
...
+
+        with torch.no_grad():
+            for parameter in src_model.parameters():
+                parameter.data.copy_(torch.randn(parameter.shape))
+
+        ema_hook.after_train_iter(runner, 1)
+        for p, p_ in zip(src_model.parameters(), ema_model.parameters()):
+            assert_allclose(p.data, p_.data)
+
+        runner.train_loop._epoch = 1
```
Comment (zhouzaida) [inline tests/test_hooks/test_ema_hook.py:139]: Why does `epoch` need to be set as 1?

File: tests/test_hooks/test_ema_hook.py [hunk 5]
```diff
@@ -62,239 +39,249 @@ class ToyModel3(ToyModel):
...
+        src_model = runner.model
+        ema_model = ema_hook.ema_model
+
+        with torch.no_grad():
+            for parameter in src_model.parameters():
+                parameter.data.copy_(torch.randn(parameter.shape))
+
+        ema_hook.after_train_iter(runner, 1)
+        for p, p_ in zip(src_model.parameters(), ema_model.parameters()):
+            assert_allclose(p.data, p_.data)
```
Comment (zhouzaida) [inline tests/test_hooks/test_ema_hook.py:137]:
```suggestion
        for src, ema in zip(src_model.parameters(), ema_model.parameters()):
            assert_allclose(src.data, ema.data)
```

File: tests/test_hooks/test_ema_hook.py [hunk 6]
```diff
@@ -62,239 +39,249 @@ class ToyModel3(ToyModel):
...
+        ema_hook.before_run(runner)
+        ema_hook.before_train(runner)
+        ema_hook.after_load_checkpoint(runner, checkpoint)
+
+        for key in checkpoint['state_dict'].keys():
+            assert_allclose(
+                checkpoint['state_dict'][key].cpu(),
+                ema_hook.ema_model.state_dict()[f'module.{key}'].cpu())
+
+        # Test a warning should be raise when resuming from a checkpoint
```
Comment (zhouzaida) [inline tests/test_hooks/test_ema_hook.py:197]:
```suggestion
        # Test a warning should be raised when resuming from a checkpoint
```

File: tests/test_hooks/test_ema_hook.py [hunk 7]
```diff
@@ -62,239 +39,249 @@ class ToyModel3(ToyModel):
...
+        # without `ema_state_dict`
+        runner._resume = True
+        ema_hook.after_load_checkpoint(runner, checkpoint)
+        with self.assertLogs(runner.logger, level='WARNING') as cm:
+            ema_hook.after_load_checkpoint(runner, checkpoint)
+            self.assertRegex(cm.records[0].msg, 'There is no `ema_state_dict`')
+
+        # Check the weight of state_dict and ema_state_dict have been swapped.
+        # when runner._resume is False
```
Comment (zhouzaida) [inline tests/test_hooks/test_ema_hook.py:206]:
```suggestion
        # when runner._resume is True
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**